README.md
# Flask-optimize

[![Build Status](https://travis-ci.org/sunary/flask-optimize.svg?branch=master)](https://travis-ci.org/sunary/flask-optimize)

**Flask optimization using cache, minify html and compress response**

https://pypi.python.org/pypi/flask-optimize

## Optimize parameters

### init_app:

**config:** Global config

Default:

```python
{
    'html': {'htmlmin': True,  'compress': True, 'cache': 'GET-84600'},
    'json': {'htmlmin': False, 'compress': True, 'cache': False},
    'text': {'htmlmin': False, 'compress': True, 'cache': 'GET-600'}
}
```

`html`, `json`, `text`: keys for data type of response, see detail below

**config_update:** update into default global config

### optimize

**dtype:** Data type of response, will get value corresponding with key in global config.

- `html` *(default)* 
- `text`
- `json`

**htmlmin:** Override `htmlmin` in config by key **dtype**

```
None: using default value from global config
False: disable
True: enable minify html
```

**compress:** Override `compress` in config by key **dtype**

```
None: using default value from global config
False: disable
True: enable compress content (using gzip)
```

**cache:** Override `cache` in config by key **dtype**

```
None: using default value from global config
False: disable
integer value: enable all method, value is cached period (seconds)
GET-84600: enable for GET method only, cached period is 84600 seconds
```


## Python code usage

```python
from flask import Flask
from flask_optimize import FlaskOptimize

flask_app = Flask(__name__)
flask_optimize = FlaskOptimize()

@flask_app.route('/')
@flask_optimize.optimize()
def index():
    return 'using Flask-optimize'

@flask_app.route('/html')
@flask_optimize.optimize()
def html():
    return '''
    <html>
    <body>
    Default data type is html.
    This content will be minified.
    </body>
    </html>
    '''
    
@flask_app.route('/text')
@flask_optimize.optimize('text')
def text():
    return '''
    <html>
    <body>
    Data type response is text, so this content wasn't minified
    </body>
    </html>
    '''
    
@flask_app.route('/json')
@flask_optimize.optimize('json')
def json():
    return {'text': 'anything', 'other_values': [1, 2, 3, 4]}

if __name__ == '__main__':
    flask_app.run('localhost', 8080, debug=True)
```

## Install

```shell
pip install flask-optimize
```

## Requirements: ##

* Python 2.x
* Python 3.x


CHANGES.md
- 2016/04/28 -- Initial project
- 2016/12/20 -- Support optimize all response
- 2018/06/24 -- Remove unnecessary code

tests/run_app.py
__author__ = 'sunary'


from flask_optimize import FlaskOptimize
from flask import Flask, request, render_template
import time


flask_app = Flask(__name__)
flask_optimize = FlaskOptimize(flask_app)


@flask_app.route('/')
@flask_optimize.optimize()
def index():
    return 'Using flask-optimize'


@flask_app.route('/html')
@flask_optimize.optimize()
def html():
    return '''
    <html>
    <body>
    The content of the body element is displayed in your browser.
    </body>
    </html>
    '''

@flask_app.route('/text')
@flask_optimize.optimize('text')
def text():
    return '''
    <html>
    <body>
    Data type response is text, so this content wasn't minified
    </body>
    </html>
    '''

@flask_app.route('/json')
@flask_optimize.optimize('json')
def json():
    return {'text': 'anything', 'other_values': [1, 2, 3, 4]}


@flask_app.route('/load_svg')
@flask_optimize.optimize()
def load_svg():
    return render_template('load_svg.html')


if __name__ == '__main__':
    flask_app.run(host='0.0.0.0', port=5372, debug=False)

    time.sleep(5)
    func = request.environ.get('werkzeug.server.shutdown')
    if func is None:
        raise RuntimeError('Not running with the Werkzeug Server')

    func()

tests/flask_app.py
__author__ = 'sunary'


from flask_optimize import FlaskOptimize
from flask import Flask, request
import time


flask_app = Flask(__name__)
flask_optimize = FlaskOptimize()


@flask_app.route('/')
@flask_optimize.optimize()
def index():
    return 'Using flask-optimize'


if __name__ == '__main__':
    flask_app.run(host='0.0.0.0', port=5372, debug=False)

    time.sleep(5)
    func = request.environ.get('werkzeug.server.shutdown')
    if func is None:
        raise RuntimeError('Not running with the Werkzeug Server')
    func()

flask_optimize.py
__author__ = 'sunary'


from htmlmin.main import minify
from flask import request, Response, make_response, current_app, redirect
from functools import update_wrapper
import gzip
import time

try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO


class FlaskOptimize(object):

    _cache = {}
    _timestamp = {}

    def __init__(self, config=None, config_update={}, redis=None):
        ''' Global config for flask optimize foreach respond return type
        Args:
            config: global configure values
            config_update: update into default configure
            redis: redis client store limit requests if you enable it
        '''
        if config is None:
            config = {'html': {'htmlmin': True, 'izip': True, 'cache': 'GET-84600'},
                      'json': {'htmlmin': False, 'izip': True, 'cache': False},
                      'text': {'htmlmin': False, 'izip': True, 'cache': 84600},
                      'limit': [100, 60, 84600],
                      'redirect_host': []}
        config.update(config_update)

        self.config = config
        self.redis = redis

    def optimize(self, type='html', htmlmin=None, izip=None, cache=None, limit=False, redirect_host=True):
        ''' Flask optimize respond using minify html, zip content and mem cache.
        Elastic optimization and create Cross-site HTTP requests if respond is json
        Args:
            type: respond return type
            htmlmin: None is using global config, True is enable minify html
            izip: None is using global config, True is enable zip respond
            cache: time cache respond. None is using global config, False or 0 to disable cache,
                integer value to set time cache (seconds),
                or string format: 'METHOD-seconds' to select METHOD cache, eg: 'GET-3600'
            limit: True if you want using default value,
                using this format [requests, window, ban expire] to set value,
                False is disable it
            redirect_host: True if you want using default value,
                using this format [['host1', 'host2], 'host_redirect'] to set value,
                False is disable it
        Examples:
            @optimize(type='html', htmlmin=True, zip=True, cache='GET-84600')
        '''

        def _decorating_wrapper(func):

            def _optimize_wrapper(*args, **kwargs):
                try:
                    load_config = self.config[type]

                    htmlmin_arg = load_config['htmlmin'] if (htmlmin is None) else htmlmin
                    izip_arg = load_config['izip'] if (izip is None) else izip
                    cache_arg = load_config['cache'] if (cache is None) else cache

                    if isinstance(cache_arg, str) and request.method in cache_arg:
                        cache_arg = int(cache_arg.split('-')[1])
                    elif not isinstance(cache_arg, int):
                        cache_arg = 0

                    limit_arg = self.config['limit'] if (limit is True) else limit
                    redirect_host_arg = self.config['redirect_host'] if (redirect_host is True) else redirect_host
                except:
                    raise 'Wrong input format'

                # limit by ip
                if limit_arg and self.redis:
                    limit_key = request.remote_addr
                    ban_key = 'ban' + limit_key

                    times_requested = self.redis.get(limit_key) or '0'
                    if int(times_requested) >= limit_arg[0] or (limit_arg[2] and self.redis.get(ban_key)):
                        if int(times_requested) >= limit_arg[0]:
                            self.redis.set(ban_key, 1)
                            self.redis.expire(ban_key, limit_arg[2])

                        return self.crossdomain({'status_code': 429})
                    elif times_requested:
                        self.redis.incr(limit_key, 1)
                    else:
                        self.redis.set(limit_key, 1)
                        self.redis.expire(limit_key, limit_arg[1])

                # redirect new host
                if redirect_host_arg:
                    for host in redirect_host_arg[0]:
                        if host in request.url_root:
                            redirect_url = request.url
                            redirect_url = redirect_url.replace(host, redirect_host_arg[1])
                            return redirect(redirect_url)

                # find cached value
                if cache_arg:
                    now = time.time()
                    key_cache = request.url

                    if self._timestamp.get(key_cache, now) > now:
                        return self._cache[key_cache]

                resp = func(*args, **kwargs)

                # crossdomain
                if type == 'json':
                    resp = self.crossdomain(resp)

                # min html
                if htmlmin_arg:
                    resp = self.validate(minify, resp)

                # gzip
                if izip_arg:
                    resp = self.validate(self.zipper, resp)

                # cache
                if cache_arg:
                    self._cache[key_cache] = resp
                    self._timestamp[key_cache] = now + cache_arg

                return resp

            return update_wrapper(_optimize_wrapper, func)

        return _decorating_wrapper

    @staticmethod
    def validate(method, content):
        if isinstance(content, (str, unicode)):
            return method(content)
        elif isinstance(content, tuple):
            return method(content[0]), content[1]

        return content

    @staticmethod
    def zipper(content):
        ''' Zip str, unicode type content
        '''
        if isinstance(content, unicode):
            content = content.encode('utf8')

        resp = Response()
        gzip_buffer = StringIO()
        gzip_file = gzip.GzipFile(mode='wb', fileobj=gzip_buffer)
        gzip_file.write(content)
        gzip_file.close()

        resp.data = gzip_buffer.getvalue()
        resp.headers['Content-Encoding'] = 'gzip'
        resp.headers['Vary'] = 'Accept-Encoding'
        resp.headers['Content-Length'] = len(resp.data)

        return resp

    @staticmethod
    def crossdomain(content):
        ''' create Cross-site HTTP requests
        '''
        resp = make_response(content)
        h = resp.headers

        h['Access-Control-Allow-Origin'] = '*'
        h['Access-Control-Allow-Methods'] = current_app.make_default_options_response().headers['allow']
        h['Access-Control-Max-Age'] = '21600'

        return resp


if __name__ == '__main__':
    flask_optimize = FlaskOptimize()
    flask_optimize.optimize('html')
    flask_optimize.optimize('json')
    flask_optimize.optimize('text')

flask_optimize/optimize.py
__author__ = 'sunary'


import sys
import gzip
import time
from htmlmin.main import minify
from flask import request, Response, make_response, current_app, json, wrappers
from functools import update_wrapper


IS_PYTHON_3 = True
if sys.version_info[0] < 3:
    from StringIO import StringIO
    IS_PYTHON_3 = False
else:
    from io import BytesIO


class FlaskOptimize(object):

    _cache = {}
    _timestamp = {}

    def __init__(self,
                 config=None):
        """
        Global config for flask optimize foreach respond return type
        Args:
            config: global configure values
        """
        if config is None:
            config = {
                        'html': {'htmlmin': True,  'compress': True, 'cache': 'GET-84600'},
                        'json': {'htmlmin': False, 'compress': True, 'cache': False},
                        'text': {'htmlmin': False, 'compress': True, 'cache': 'GET-84600'},
                        'trim_fragment': False,
                     }

        self.config = config

    def optimize(self,
                 dtype='html',
                 htmlmin=None,
                 compress=None,
                 cache=None):
        """
        Flask optimize respond using minify html, zip content and mem cache.
        Elastic optimization and create Cross-site HTTP requests if respond is json
        Args:
            dtype: response type:
                - `html` (default)
                - `text`
                - `json`
            htmlmin: minify html
                None (default): using global config,
                False: disable minify html
                True: enable minify html
            compress: send content in compress (gzip) format
                None (default): using global config,
                False: disable compress response,
                True: enable compress response
            cache: cache content in RAM
                None (default): using global config,
                False: disable cache,
                integer: cache all method with period
                string value: 'METHOD-seconds' to select METHOD and period cache, eg: 'GET-3600', 'GET|POST-600', ...
        Examples:
            @optimize(dtype='html', htmlmin=True, compress=True, cache='GET-84600')
        """

        def _decorating_wrapper(func):

            def _optimize_wrapper(*args, **kwargs):
                # default values:
                is_htmlmin = False
                is_compress = False
                period_cache = 0

                if self.config.get(dtype):
                    is_htmlmin = self.config.get(dtype)['htmlmin'] if (htmlmin is None) else htmlmin
                    is_compress = self.config.get(dtype)['compress'] if (compress is None) else compress
                    cache_agrs = self.config.get(dtype)['cache'] if (cache is None) else cache

                    if cache is False or cache == 0:
                        period_cache = 0
                    elif isinstance(cache_agrs, int):
                        period_cache = cache_agrs
                    elif isinstance(cache_agrs, str if IS_PYTHON_3 else (str, basestring)) and\
                            len(cache_agrs.split('-')) == 2:
                        try:
                            period_cache = int(cache_agrs.split('-')[1]) if (request.method in cache_agrs) else 0
                        except (KeyError, ValueError):
                            raise ValueError('Cache must be string with method and period cache split by "-"')
                    else:
                        raise ValueError('Cache must be False, int or string with method and period cache split by "-"')

                # init cached data
                now = time.time()
                key_cache = request.url
                if self.config.get('trim_fragment'):
                    key_cache = key_cache.split('#')[0]

                if self._timestamp.get(key_cache) > now:
                    return self._cache[key_cache]

                resp = func(*args, **kwargs)

                if not isinstance(resp, wrappers.Response):
                    # crossdomain
                    if dtype == 'json':
                        resp = self.crossdomain(resp)

                    # min html
                    if is_htmlmin:
                        resp = self.validate(self.minifier, resp)

                    # compress
                    if is_compress:
                        resp = self.validate(self.compress, resp)

                # cache
                if period_cache > 0:
                    self._cache[key_cache] = resp
                    self._timestamp[key_cache] = now + period_cache

                return resp

            return update_wrapper(_optimize_wrapper, func)

        return _decorating_wrapper

    @staticmethod
    def validate(method, content):
        instances_compare = (str, Response) if IS_PYTHON_3 else (str, unicode, Response)
        if isinstance(content, instances_compare):
            return method(content)
        elif isinstance(content, tuple):
            if len(content) < 2:
                raise TypeError('Content must have larger than 2 elements')

            return method(content[0]), content[1]

        return content

    @staticmethod
    def minifier(content):
        if not IS_PYTHON_3 and isinstance(content, str):
            content = unicode(content, 'utf-8')

        return minify(content,
                      remove_comments=True,
                      reduce_empty_attributes=True,
                      remove_optional_attribute_quotes=False)

    @staticmethod
    def compress(content):
        """
        Compress str, unicode content using gzip
        """
        resp = Response()
        if isinstance(content, Response):
            resp = content
            content = resp.data

        if not IS_PYTHON_3 and isinstance(content, unicode):
            content = content.encode('utf8')

        if IS_PYTHON_3:
            gzip_buffer = BytesIO()
            gzip_file = gzip.GzipFile(fileobj=gzip_buffer, mode='wb')
            gzip_file.write(bytes(content, 'utf-8'))
        else:
            gzip_buffer = StringIO()
            gzip_file = gzip.GzipFile(fileobj=gzip_buffer, mode='wb')
            gzip_file.write(content)

        gzip_file.close()

        resp.data = gzip_buffer.getvalue()
        resp.headers['Content-Encoding'] = 'gzip'
        resp.headers['Vary'] = 'Accept-Encoding'
        resp.headers['Content-Length'] = len(resp.data)

        return resp

    @staticmethod
    def crossdomain(content):
        """
        Create decorator Cross-site HTTP requests
        see more at: http://flask.pocoo.org/snippets/56/
        """
        if isinstance(content, (dict, Response)):
            if isinstance(content, dict):
                content = json.jsonify(content)
                resp = make_response(content)
            elif isinstance(content, Response):
                resp = content

            h = resp.headers
            h['Access-Control-Allow-Origin'] = '*'
            h['Access-Control-Allow-Methods'] = current_app.make_default_options_response().headers['allow']
            h['Access-Control-Max-Age'] = '21600'

            return resp

        return content


if __name__ == '__main__':
    flask_optimize = FlaskOptimize()
    flask_optimize.optimize('html')
    flask_optimize.optimize('json')
    flask_optimize.optimize('text')
